File: src/lib/dashscope.ts
import { DASHSCOPE_API_KEY } from './secrets';
import { logAiUsage, detectFeature } from './aiUsage';

const DASHSCOPE_BASE_URL = 'https://dashscope.aliyuncs.com/compatible-mode/v1';
const DASHSCOPE_NATIVE_URL = 'https://dashscope.aliyuncs.com/api/v1/services/aigc/multimodal-generation/generation';
const QWEN_VL_MODEL = 'qwen3-vl-plus';
const QWEN_IMAGE_MODEL = 'qwen-image-2.0-pro';

export interface DashScopeMessage {
  role: 'system' | 'user' | 'assistant';
  content: string | DashScopeContentBlock[];
}

export interface DashScopeContentBlock {
  type: 'text' | 'image_url';
  text?: string;
  image_url?: { url: string };
}

export interface DashScopeOptions {
  model?: string;
  temperature?: number;
  maxTokens?: number;
  jsonMode?: boolean;
}

function isAvailable(): boolean {
  return DASHSCOPE_API_KEY.length > 0;
}

export async function qwenVisionChat(
  imageUri: string,
  prompt: string,
  options?: DashScopeOptions,
): Promise<string | null> {
  if (!isAvailable()) {
    console.warn('[DashScope] API Key not set, skipping vision call');
    return null;
  }

  const t0 = Date.now();
  const model = options?.model || QWEN_VL_MODEL;
  const feature = detectFeature(prompt);

  let imageUrl: string;
  if (imageUri.startsWith('data:')) {
    imageUrl = imageUri;
  } else if (imageUri.startsWith('http')) {
    imageUrl = imageUri;
  } else {
    try {
      const response = await fetch(imageUri);
      const blob = await response.blob();
      const base64 = await new Promise<string>((resolve) => {
        const reader = new FileReader();
        reader.onloadend = () => {
          const result = reader.result as string;
          const base64Data = result.split(',')[1] || result;
          resolve(base64Data);
        };
        reader.readAsDataURL(blob);
      });
      const mime = blob.type || 'image/jpeg';
      imageUrl = `data:${mime};base64,${base64}`;
    } catch (e) {
      console.warn('[DashScope] Failed to read image file:', e);
      return null;
    }
  }

  const messages: DashScopeMessage[] = [
    {
      role: 'user',
      content: [
        { type: 'image_url', image_url: { url: imageUrl } },
        { type: 'text', text: prompt },
      ],
    },
  ];

  try {
    const body: Record<string, unknown> = {
      model: options?.model || QWEN_VL_MODEL,
      messages,
      stream: false,
      temperature: options?.temperature ?? 0.3,
      max_tokens: options?.maxTokens ?? 2048,
    };
    if (options?.jsonMode) {
      body.response_format = { type: 'json_object' };
    }

    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), 30000);

    const res = await fetch(`${DASHSCOPE_BASE_URL}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${DASHSCOPE_API_KEY}`,
      },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    clearTimeout(timer);

    if (!res.ok) {
      console.warn('[DashScope] Vision API error:', res.status, await res.text().catch(() => ''));
      logAiUsage({ provider: 'qwen', model, feature, callType: 'vision', durationMs: Date.now() - t0, ok: false });
      return null;
    }

    const data = await res.json();
    const u = data.usage ?? {};
    const content = data.choices?.[0]?.message?.content;
    const ok = typeof content === 'string' && !!content.trim();
    logAiUsage({
      provider: 'qwen', model, feature, callType: 'vision',
      promptTokens: u.prompt_tokens, cachedTokens: u.prompt_tokens_details?.cached_tokens,
      completionTokens: u.completion_tokens, requestId: data.id,
      durationMs: Date.now() - t0, ok,
    });
    if (!ok) return null;
    return content.trim();
  } catch (e) {
    console.warn('[DashScope] Vision request failed:', e);
    logAiUsage({ provider: 'qwen', model, feature, callType: 'vision', durationMs: Date.now() - t0, ok: false });
    return null;
  }
}

/**
 * qwen-image-2.0-pro must use DashScope's native MultiModalConversation endpoint,
 * NOT the OpenAI-compatible-mode endpoint (which returns empty content).
 *
 * Native endpoint: POST /api/v1/services/aigc/multimodal-generation/generation
 * Response: output.choices[0].message.content[0].image = "https://..."
 */
export async function qwenGenerateImage(
  prompt: string,
  options?: { size?: string; imageUrl?: string; refImage?: string },
): Promise<string | null> {
  if (!isAvailable()) {
    console.warn('[DashScope] API Key not set, skipping image generation');
    return null;
  }

  const t0 = Date.now();
  const feature = detectFeature(prompt);

  try {
    // Build content parts: optional reference image + text prompt
    const contentParts: { image?: string; text?: string }[] = [];

    if (options?.refImage) {
      // Convert local/file URI to base64 data URL if needed
      let refUrl = options.refImage;
      if (!refUrl.startsWith('data:') && !refUrl.startsWith('http')) {
        try {
          const response = await fetch(refUrl);
          const blob = await response.blob();
          const base64 = await new Promise<string>((resolve) => {
            const reader = new FileReader();
            reader.onloadend = () => {
              const result = reader.result as string;
              resolve(result);
            };
            reader.readAsDataURL(blob);
          });
          refUrl = base64;
        } catch (e) {
          console.warn('[DashScope] Failed to read ref image:', e);
        }
      }
      if (refUrl.startsWith('data:') || refUrl.startsWith('http')) {
        contentParts.push({ image: refUrl });
      }
    }

    contentParts.push({ text: prompt });

    const body = {
      model: QWEN_IMAGE_MODEL,
      input: {
        messages: [
          {
            role: 'user',
            content: contentParts,
          },
        ],
      },
    };

    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), 120000);

    const res = await fetch(DASHSCOPE_NATIVE_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${DASHSCOPE_API_KEY}`,
      },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    clearTimeout(timer);

    if (!res.ok) {
      console.warn('[DashScope] Image gen error:', res.status, await res.text().catch(() => ''));
      logAiUsage({ provider: 'qwen', model: QWEN_IMAGE_MODEL, feature, callType: 'image', durationMs: Date.now() - t0, ok: false });
      return null;
    }

    const data = await res.json();
    const imageCount = data.usage?.image_count ?? undefined;

    // Native format: output.choices[0].message.content[0].image
    const content = data.output?.choices?.[0]?.message?.content;
    if (Array.isArray(content)) {
      for (const item of content) {
        if (typeof item === 'object' && item.image && typeof item.image === 'string') {
          logAiUsage({ provider: 'qwen', model: QWEN_IMAGE_MODEL, feature, callType: 'image', imageCount: imageCount ?? 1, requestId: data.request_id, durationMs: Date.now() - t0, ok: true });
          return item.image;
        }
      }
    }

    console.warn('[DashScope] Image gen: unexpected response', JSON.stringify(data).slice(0, 300));
    logAiUsage({ provider: 'qwen', model: QWEN_IMAGE_MODEL, feature, callType: 'image', imageCount, requestId: data.request_id, durationMs: Date.now() - t0, ok: false });
    return null;
  } catch (e) {
    console.warn('[DashScope] Image gen failed:', e);
    logAiUsage({ provider: 'qwen', model: QWEN_IMAGE_MODEL, feature, callType: 'image', durationMs: Date.now() - t0, ok: false });
    return null;
  }
}

export { isAvailable };

File: model-service/stylee/vision/dashscope.py
"""真 provider:识别/回验 qwen3-vl-plus(OpenAI-compatible chat,含图);
标准化 qwen-image-edit(图生图/img2img,multimodal-generation 同步端点,base64 进、图 URL 出)。

纯 urllib,自动走环境代理。识别复用 providers.openai_compat 的 _chat_completion。
注:qwen-image-edit 返回的 OSS URL 会过期,落库由调用方负责。
图像模型选型约定:图生图/图片编辑 → qwen-image-edit;文生图(如从零合成) → qwen-image。
标准化两种模式(cutout/img2img)都带输入图,故统一走 qwen-image-edit。
"""
from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request

from ..usage_log import log_usage

from ..contracts import WardrobeItem
from ..providers.openai_compat import _chat_completion
from . import prompts
from .base import ImageStandardizer, VisionProvider
from .mock import MockImageStandardizer, MockVisionProvider


class VisionError(RuntimeError):
    pass


_EDIT_PROMPTS = {
    "cutout": "去掉背景,把这件衣物单独放到纯白背景上,平铺商品图,保留原本的颜色、材质和图案。",
    "img2img": "以这张图里的衣物为准,重绘成干净的纯白背景平铺商品图,保留颜色、材质、版型和图案,不要人物。",
}

_MM_URL_PATH = "/services/aigc/multimodal-generation/generation"


# --- 标准化:请求/响应纯函数(可离线测) ---
def build_edit_payload(model: str, image_url: str, prompt: str) -> dict:
    return {"model": model, "input": {"messages": [{"role": "user", "content": [
        {"image": image_url}, {"text": prompt}]}]}, "parameters": {}}


def parse_edit_response(body: dict) -> str:
    try:
        return body["output"]["choices"][0]["message"]["content"][0]["image"]
    except (KeyError, IndexError, TypeError):
        raise VisionError(f"标准化响应无图: {str(body)[:200]}") from None


class DashScopeVisionProvider(VisionProvider):
    def __init__(self, base_url: str, api_key: str, model: str, timeout: int = 60):
        if not api_key:
            raise VisionError("DashScopeVisionProvider: 缺少 api_key(DASHSCOPE_API_KEY)")
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.model = model
        self.name = model
        self.timeout = timeout

    def recognize(self, image_url: str) -> dict:
        content = _chat_completion(self.base_url, self.api_key, self.model,
                                   prompts.build_recognize_messages(image_url),
                                   0.2, self.timeout, True)
        return prompts.parse_recognize_json(content)

    def verify(self, image_url: str, expected: dict) -> dict:
        content = _chat_completion(self.base_url, self.api_key, self.model,
                                   prompts.build_verify_messages(image_url, expected),
                                   0.0, self.timeout, True)
        return prompts.parse_verify_json(content)


class DashScopeImageStandardizer(ImageStandardizer):
    def __init__(self, base_url: str, api_key: str, model: str, timeout: int = 60):
        if not api_key:
            raise VisionError("DashScopeImageStandardizer: 缺少 api_key")
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.model = model
        self.name = model
        self.timeout = timeout

    def standardize(self, image_url: str, item: WardrobeItem, mode: str) -> str:
        prompt = _EDIT_PROMPTS.get(mode, _EDIT_PROMPTS["img2img"])
        data = json.dumps(build_edit_payload(self.model, image_url, prompt)).encode("utf-8")
        req = urllib.request.Request(
            self.base_url + _MM_URL_PATH, data=data, method="POST",
            headers={"Content-Type": "application/json",
                     "Authorization": f"Bearer {self.api_key}"})
        t0 = time.time()
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                body = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            log_usage("qwen", self.model, "standardize", "image", None, int((time.time() - t0) * 1000), False)
            raise VisionError(f"HTTP {e.code}: {e.read().decode('utf-8','replace')[:200]}") from None
        except urllib.error.URLError as e:
            log_usage("qwen", self.model, "standardize", "image", None, int((time.time() - t0) * 1000), False)
            raise VisionError(f"网络错误: {e.reason}") from None
        log_usage("qwen", self.model, "standardize", "image", body.get("usage"),
                  int((time.time() - t0) * 1000), True, body.get("request_id"))
        return parse_edit_response(body)


_DASHSCOPE = "https://dashscope.aliyuncs.com"


def build_vision_provider() -> VisionProvider:
    key = os.environ.get("DASHSCOPE_API_KEY", "")
    if not key:
        print("[vision] 无 key → MockVisionProvider 降级")
        return MockVisionProvider()
    return DashScopeVisionProvider(
        base_url=os.environ.get("VL_BASE_URL", _DASHSCOPE + "/compatible-mode/v1"),
        api_key=key, model=os.environ.get("VL_MODEL", "qwen3-vl-plus"))


def build_image_standardizer() -> ImageStandardizer:
    key = os.environ.get("DASHSCOPE_API_KEY", "")
    if not key:
        print("[vision] 无 key → MockImageStandardizer 降级")
        return MockImageStandardizer()
    return DashScopeImageStandardizer(
        base_url=os.environ.get("IMG_BASE_URL", _DASHSCOPE + "/api/v1"),
        api_key=key, model=os.environ.get("IMG_EDIT_MODEL", "qwen-image-edit"))
